### ISSUE TITLE
Should hook up gc

### ISSUE THREAD

**Issue by ranjanan** [JuliaComputing/ArrayFire.jl#27]
I should hook up `AFArrays` to the finalizer.

**Comment by pavanky** [comment]
I haven't used Julia in many many years, but is the finalizer run when the variable goes out of scope or is it run non a deterministic manner ?

**Comment by Keno** [comment]
It's non-deterministic.

**Comment by pavanky** [comment]
If it is non-deterministic the GPU can run out of memory very very easily.
A simple test would be to run a matrix multiply of [5000, 5000] in a loop for a thousand times. 

If there is a deterministic way to call the finalizers, you can implement something similar to [what I did with R](https://github.com/arrayfire/arrayfire-r/blob/master/R/array.R#L4).

But when I [last came across this issue](https://groups.google.com/forum/#!topic/julia-dev/B3g6lLViOjo), Julia did not have a function to call finalizers deterministically either.

**Comment by ViralBShah** [comment]
This issue comes up in distributed computing too. For now, we should have the finalizers and eventually adopt any julia solution that might come up.

**Comment by ranjanan** [comment]
I added a simple finalizer here: 79e39e92cc85fb9f4bc5af7ce5ff391f546a0874

But when I use the CPU backend, I'm still seeing substantial use of swap space if I run a reasonable application again and again. Though `top` seems to be stable, disk space seems to be running out on my Mac.

**Comment by pavanky** [comment]
I looked through the old thread again, and it looks like `gc()` has been present all along.

Can't you just make `af_device_mem_info` part of `_AFArray` and call `gc()` when you hit a limit ? You can have a default value for this and let the user set it so it will not affect the overall performance.

**Comment by ranjanan** [comment]
The following code snippet: 
```julia
using ArrayFire
setBackend(AF_BACKEND_CPU)
function thing()
    a = rand(AFArray{Float64}, 10, 10)
    b = a + 1
    c = a + 2
    @show deviceMemInfo()
    nothing
end
thing()
thing()
thing()
println("Final:", deviceMemInfo())
```
gives the following output:
```
deviceMemInfo() = (1024,1,1024,1)
deviceMemInfo() = (2048,2,2048,2)
deviceMemInfo() = (3072,3,3072,3)
Final:(3072,3,3072,3)
```

**Comment by gabgoh** [comment]
I think the absence of a garbage collector makes this library unusable. I've been using a 3rd party program to track my GPU memory usage, and the NNF benchmark takes up about 1.5 G of ram, none of which are freed. Running the benchmark 3 times in the row results in an outofmemory error.

**Comment by ranjanan** [comment]
@gabgoh I pushed some fixes, and it might be slightly better. `gc()` and `af_device_gc()` work now. But you might have call one of those functions from time to time. Looking into ways to automate this.

**Comment by gabgoh** [comment]
Whoa, thanks for the fast response! I was just testing this out, and can confirm it works beautifully!!

To anyone else experiencing this issue, just call gc() and then ArrayFire.af_device_gc() inside any loop which allocates memory. This fixes the memory leak.

**Comment by ViralBShah** [comment]
@carnaval @JeffBezanson @Keno @StefanKarpinski Is there a way to do this better?

**Comment by Keno** [comment]
When we discussed this, Jeff was proposing adding memory pressure proportional to the ratio of main memory to GPU memory to trick the GC into feeling the memory pressure. Longer term the memory use can tracked by separate heap accounting for CPU and GPU.

**Comment by pavanky** [comment]
Right now arrayfire does not expose the total memory of the GPU. This might change in the future to help this out.

@gabgoh with the finalizer in place, you should not need to call `af_device_gc()`.

**Comment by gabgoh** [comment]
@pavanky 

Here is a snippet of code I just ran

```
julia> A = randn(AFArray{Float64}, 5000, 5000);

julia> A = 2
2

julia> gc();

julia> deviceMemInfo()
(200000512,1,0,0)

julia> gc();

julia> deviceMemInfo()
(200000512,1,0,0)

julia> af_device_gc()
ERROR: UndefVarError: af_device_gc not defined

julia> ArrayFire.af_device_gc()
true

julia> deviceMemInfo()
(0,0,0,0)
```

my third party tool informs me that the memory is only really cleared after af_device_gc() has been called, so I'm presuming arrayfire's own garbage collector has been notified but not run.

**Comment by pavanky** [comment]
@gabgoh the last two 0s are important. The first two get reused if you create an array again.

**Comment by gabgoh** [comment]
makes sense!

**Comment by pavanky** [comment]
here's the order for reference:

allocated bytes, allocated buffers, locked bytes, locked buffers.

allocated means arrayfire memory manager has created a memory pool of that size. The lock bytes / buffers signify how many arrays are still referenced and / or not finalized. 

The memory in the allocated pool is reused for buffers of similar size. you need not call `af_device_gc`.

**Comment by ViralBShah** [comment]
Cc @yuyichao @vchuravy and @maleadt in case he has some good ideas on how to detect GPU memory pressure.

**Comment by pavanky** [comment]
I don't know how easy this would be (I haven't seen this done anywhere), but if a user can define memory pressure for any given class and tie it to the existing gc (or a new instance of it), it would be a great abstraction instead of having special code inside of Julia to detect memory pressure for various devices.

**Comment by gaika** [comment]
In https://github.com/JuliaComputing/ArrayFire.jl/pull/121 I have `@afgc` macro that tracks creation of the temporary arrays in a function (or during assignment) to releases them immediately without waiting for `gc()`. That significantly reduces the pressure when there are a lots of temporary arrays. 

Separately GPU memory pressure is detected on every call and if above some threshold Julia's `gc()` is called automatically. Would work even better when https://github.com/arrayfire/arrayfire/issues/1704 is fixed.

**Comment by gaika** [comment]
merged with [v0.1.0](https://github.com/JuliaComputing/ArrayFire.jl/releases/tag/v0.1.0)
